src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -0,0 +1,282 @@
+        where T : IShapeable<T>
+    {
+        MessagePackReader reader = new(sequence);
+        return SerializerCache<T>.Deserializer(ref reader);
+    }
+
+    private delegate T? Reader<T>(ref MessagePackReader reader);
+
+    private delegate void Writer<T>(ref MessagePackWriter writer, T value);
+
+    private static class SerializerCache<T>
+        where T : IShapeable<T>
+    {
+        private static Writer<T?>? serializer;
+
+        private static Reader<T?>? deserializer;
+
+        internal static Writer<T?> Serializer => serializer ??= SerializeVisitor.GetWriter(T.GetShape());
+
+        internal static Reader<T?> Deserializer => deserializer ??= DeserializeVisitor.GetReader(T.GetShape());
+    }
+
+    private sealed class SerializeVisitor : TypeShapeVisitor
+    {
+        internal static readonly SerializeVisitor Instance = new();
+
+        private static readonly TypeDictionary formatters = new()
+        {
+            { typeof(string), new Writer<string>((ref MessagePackWriter writer, string value) => writer.Write(value)) },
+            { typeof(int), new Writer<int>((ref MessagePackWriter writer, int value) => writer.Write(value)) }
+        };
+
+        internal static Writer<T?> GetWriter<T>(ITypeShape<T> typeShape)
+        {
+            return formatters.GetOrAdd<Writer<T?>>(typeShape, Instance, box => (ref MessagePackWriter writer, T? value) => box.Result(ref writer, value));
+        }
+
+        private SerializeVisitor() { }
+
+        public override object? VisitObject<T>(IObjectTypeShape<T> objectShape, object? state = null)
+        {
+            Dictionary<ReadOnlyMemory<byte>, Writer<T>> properties = new();
+            foreach (IPropertyShape property in objectShape.GetProperties())
+            {
+                if (property.HasGetter/* && property.HasSetter*/)
+                {
+                    byte[] propertyNameBytes = MessagePackSerializer.Serialize(property.Name, MessagePackSerializerOptions.Standard);
+                    properties[propertyNameBytes] = (Writer<T>)property.Accept(this, state)!;
+                }
+            }
+
+            return new Writer<T?>((ref MessagePackWriter writer, T? value) =>
+            {
+                if (value is null)
+                {
+                    writer.WriteNil();
+                    return;
+                }
+
+                writer.WriteMapHeader(properties.Count);
+
+                foreach (var property in properties)
+                {
+                    writer.WriteRaw(property.Key.Span);
+                    property.Value(ref writer, value);
+                }
+            });
+        }
+
+        public override object? VisitProperty<TDeclaringType, TPropertyType>(IPropertyShape<TDeclaringType, TPropertyType> propertyShape, object? state = null)
+        {
+            var getter = propertyShape.GetGetter();
+
+            switch (propertyShape.PropertyType.Kind)
+            {
+                case TypeShapeKind.Object:
```

Comment:
I'm guessing that this switch statement wouldn't be necessary assuming that you're returning a writer for all shape kinds.